ssh/tailssh/tailssh.go

```diff
@@ -184,46 +178,106 @@ func (c *conn) logf(format string, args ...any) {
 // PublicKeyHandler implements ssh.PublicKeyHandler is called by the
 // ssh.Server when the client presents a public key.
 func (c *conn) PublicKeyHandler(ctx ssh.Context, pubKey ssh.PublicKey) error {
-	c.mu.Lock()
-	ci := c.info
-	c.mu.Unlock()
-	if ci == nil {
+	if c.info == nil {
 		return gossh.ErrDenied
 	}
 
-	if err := c.checkAuth(pubKey); err != nil {
+	if err := c.doPolicyAuth(ctx, pubKey); err != nil {
 		// TODO(maisem/bradfitz): surface the error here.
 		c.logf("rejecting SSH public key %s: %v", bytes.TrimSpace(gossh.MarshalAuthorizedKey(pubKey)), err)
 		return err
 	}
+	if err := c.isAuthorized(ctx); err != nil {
+		return err
+	}
 	c.logf("accepting SSH public key %s", bytes.TrimSpace(gossh.MarshalAuthorizedKey(pubKey)))
 	return nil
 }
 
+// isAuthorized returns nil if the connection is authorized to proceed.
+func (c *conn) isAuthorized(ctx ssh.Context) error {
+	action := c.currentAction
+	for {
+		if action.Accept {
+			if c.pubKey != nil {
+				metricPublicKeyAccepts.Add(1)
+			}
+			return nil
+		}
+		if action.Reject || action.HoldAndDelegate == "" {
+			return gossh.ErrDenied
+		}
+		var err error
+		action, err = c.resolveNextAction(ctx)
+		if err != nil {
+			return err
+		}
+	}
+}
+
 // errPubKeyRequired is returned by NoClientAuthCallback to make the client
 // resort to public-key auth; not user visible.
 var errPubKeyRequired = errors.New("ssh publickey required")
 
+// BannerCallback implements ssh.BannerCallback.
+// It is responsible for starting the policy evaluation, and returns
+// the first message found in the action chain. It stops the evaluation
+// on the first "accept" or "reject" action, and returns the message
+// associated with that action (if any).
+func (c *conn) BannerCallback(ctx ssh.Context) string {
+	if err := c.setInfo(ctx); err != nil {
+		c.logf("failed to get conninfo: %v", err)
+		return gossh.ErrDenied.Error()
+	}
+	if err := c.doPolicyAuth(ctx, nil /* no pub key */); err != nil {
+		return ""
+	}
+	action := c.currentAction
+	for {
+		if action.Reject || action.Accept || action.Message != "" {
+			return action.Message
```

Comment:
oh, right, this is part of why we did it in stderr before instead of banner: the whole NextAction stream was made assuming that there could be N actions in a row that all emitted a banner.

The BannerCallback only lets us do one. I guess that works for now since we're not currently ever sending more than 1 from the server.

Reply:
So I think the constraint that we can only show one message to the user is fine.

Specifically this change makes Tailscale SSH be more compliant with the protocol as we do actually do auth in the auth phase now

This change also means that things like LocalPortForwarding no longer need to have a shell started. The fix in https://github.com/tailscale/tailscale/pull/5858 is very specific to LocalPortForwarding, this is fixes the underlying problem instead.